### ISSUE TITLE
Credit Card number parameter

### ISSUE THREAD

**Issue by import-brain** [faker-js/faker#181]
<img width="746" alt="Screen Shot 2022-01-16 at 2 35 07 PM" src="https://user-images.githubusercontent.com/68196762/149675063-9f980495-3ee4-498a-9fd1-fe7f7ef76d22.png">

In the part of the Finance section of the API docs shown above, I think it would improve the docs if we listed all possible inputs of the 'provider' parameter, such as 'Visa', 'MasterCard', etc. What would happen if someone used gibberish as a parameter there, would an error be thrown?

**Comment by clarkerican** [comment]
One interesting challenge here is that each locale is going to have different options. We should consider the best way to represent this, cause I think adding the options is likely a good call! But we should probably add the options for each locale.

**Comment by import-brain** [comment]
> One interesting challenge here is that each locale is going to have different options. We should consider the best way to represent this, cause I think adding the options is likely a good call! But we should probably add the options for each locale.

Yes, we probably need to look at different ways to represent this, because this section taking up half of the page is not optimal.

**Comment by ageddesi** [comment]
What if you provide the options with a dropdown for each local?

**Comment by ST-DDT** [comment]
IMO we should describe a way to look these up instead.

**Comment by xDivisionByZerox** [comment]
**Team Decision**

We came to a conclusion regarding the acceptance criteria for the implementation of this issue:

- the casing of enum values must be "kebab-case"
- soft introduction of enum/literal union values via the [`LiteralUnion`](https://github.com/faker-js/faker/blob/048c32581bd5d15f333ef63e4ca968e02793373d/src/internal/types.ts#L6-L8) helper type 
- deprecate the custom format (separate issue - we already wanted to do that)
  - extra signature for custom formats (example two signatures, FIRST one with `${string}#${string}`, other is just `LiteralUnion` of the then existing enum values)
  - runtime deprecation for `else if` case (custom format)

**Comment by UlikGames** [comment]
Sooo here’s the deal.

The acceptance criteria say kebab-case, right? But our locale keys are actually snake_case - think `american_express`. Then we go `issuer.toLowerCase() in localeFormat`, which means `'american-express'` just falls flat and matches nada. Instead of crashing, it grabs some random issuer by accident. Sneaky bug - no warnings, just wrong.

My fix, at least how I see it: bring in a `LiteralUnion` with kebab-case values and swapped hyphens for underscores when looking up. Now both formats play nice together.

One snag, though: `creditCardIssuer()` still pumps out snake_case. Should we fix that now and break things, or save the chaos for another day?

**Comment by ST-DDT** [comment]
This may be/involve a breaking change.

- CCIssuer returns Enum (locale filtered) or LiteralUnionString
- CCNumber deprecate legacy string parameter for 1 major version (implicitly filters the CCIssuers by locale/may use resolver functions)
  -  #3176
- new CCNumber method varaiants for each well known issuer (similar to the image providers/methods)

**Comment by UlikGames** [comment]
Totally makes sense. Down to split it up?

So here's the deal: the no-break tweak is basically just tossing `LiteralUnion` onto the `issuer` param and smoothing out hyphens when we look stuff up. That pumps the issuer values straight into the types and docs, exactly what this ticket was asking for, and nobody’s code breaks if it already works today. Only real change? `'american-express'` finally gives you a legit Amex number instead of pulling a random one, which, surprise, it was actually bugged before.

Only heads-up: autocomplete suggestions will be in kebab-case, but `creditCardIssuer()` still spits out snake_case for now. That mismatch sticks around till your enum lands and cleans it up. I’m cool with that, if you are.

As for switching to the enum return, deprecating the old param, and those per-issuer methods? The enum's the breaking change (so major version), but the other two are safe adds whenever you're ready.

I’m good pushing the first bit now, or holding back if you’d rather drop it all as one clean package. Up to you.

**Comment by matthewmayer** [comment]
Thanks for the update! 

I have a small suggestion: if you’re using AI to help draft replies, could you instruct it to drop the chatty 'personality' and filler (like "So here's the deal" or "which, surprise" )?

All the conversational text makes it hard to quickly scan for the core technical details during review. Direct bullet points on changes, side effects, and next steps work best for our workflow and will help get changes discussed and approved more quickly.

**Comment by UlikGames** [comment]
I got it.

Turkish is my first language, so I draft literally and lean on an AI to clean up the English and grammar. That's where the filler came from. My mistake but if you translate it Turkish you gonna see, it make sense and that's not filler (at least in Turkish). But I'm gonna review everything for next posts.

Back to the open question: is the `LiteralUnion`-only slice worth shipping given the locale gap, or should it wait for the locale-filtered enum?

**Comment by ST-DDT** [comment]
I'm not sure what the best way forward is here.

IMO we should align this with the imageProviders plus underlying locale data support:

**anytime/non-breaking**:

- we should start splitting the creditCardNumber method per issuer (`CreditCardIssuer.FOO` -> creditCardIssuerFoo())
- Then deprecate the format/issuer parameter entirely

**v11/breaking**:

- change definitions to `Record<LiteralUnion<CreditCardIssuer(Type)>, SMF/(fakerCore) => string>`
  aka `{ [CreditCardIssuer.FOO]: imported creditCardIssuerFoo method, ... }`
  - creditCardIssuer => keys()
  - creditCardNunber => values()(fakerCore)
  - Do we need the ability to generate these independent of locale?
    (aka base locale => all issuers, other locales => only issuers available there, or similar)

**Comment by UlikGames** [comment]
I would keep issuer-specific generation seperate from locale

for the non-breaking change, expose methods like `creditCardNumberVisa()` (and others) and route the current issuer parameter through them.

Deprecate the issuer parameter once you have a method for each supported issuer. Custom formats could point the user to `replaceCreditCardSymbols()`.

for v11, have a single resolver record, which `creditCardIssuer()` can pick based on the locale issuer preferences, and have `creditCardNumber()` call the appropriate resolver. The base locale would define all the issuers, as a fallback.

I would not put resolver maps in each locale as the same patterns already apply to Visa and Mastercard in multiple locales. This approach would unnecessarily bloat the library, and having the same patterns in several places makes it harder to keep them in sync. Having the same issuers in multiple locales also makes it hard to have an explicit method for an issuer, because an explicit method would be coupled to a specific locale.

Edit: Also #2720 has the same separation, with patterns being global and issuers defined per locale.

**Comment by ST-DDT** [comment]
I proposed an issuer map to allow for user provided ones to be added. They are not new implememtations, just references to the existing functions. But this usecase may not be needed anyway.

**Comment by ST-DDT** [comment]
> Custom formats could point the user to `replaceCreditCardSymbols()`.

Yes, but maybe we should give that method a better name. Aka creditCardNumberFromPattern or similar.
That way it would follow the same naming pattern as creditCardNumberVisa. Depends a bit on where it is used throughout the code.

**Comment by UlikGames** [comment]
> I proposed an issuer map to allow for user provided ones to be added. They are not new implememtations, just references to the existing functions. But this usecase may not be needed anyway.

For user-provided issuers, I'd keep the map out unless there's a specific use case. Users can call their own generator directly, and adding it to the definitions creates another public API surface to maintain.



> Yes, but maybe we should give that method a better name. Aka creditCardNumberFromPattern or similar. That way it would follow the same naming pattern as creditCardNumberVisa. Depends a bit on where it is used throughout the code.

I checked where `replaceCreditCardSymbols()` is used. `phone.imei()` uses it too, so `creditCardNumberFromPattern()` would be misleading there. Maybe `fromLuhnPattern()` would fit both? Or we could leave the generic helper as-is and add `finance.creditCardNumberFromPattern()` as a convenience method.
